delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/IdResolution.scala

```diff
@@ -0,0 +1,152 @@
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.jsonld.JsonLdContent
+import ch.epfl.bluebrain.nexus.delta.sdk.model.search.SearchResults.searchResultsJsonLdEncoder
+import ch.epfl.bluebrain.nexus.delta.sdk.model.search.{SearchResults, SortList}
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.{Label, ProjectRef, ResourceRef}
+import io.circe.{Encoder, Json, JsonObject}
+import monix.bio.{IO, UIO}
+
+/**
+  * @param defaultViewsQuery
+  *   how to list resources from the default elasticsearch views
+  * @param fetchResource
+  *   how to fetch a resource given a resourceRef and the project it lives in
+  */
+class IdResolution(
+    defaultViewsQuery: DefaultViewsQuery.Elasticsearch,
+    fetchResource: (ResourceRef, ProjectRef) => UIO[Option[JsonLdContent[_, _]]]
+) {
+
+  /**
+    * Attempts to resolve the provided identifier across projects that the caller has access to
+    *
+    *   - If the query response is empty, leads to AuthorizationFailed.
+    *   - If the query returns a single result, attempts to fetch the resource
+    *   - If there are multiple results, they are returned as [[SearchResults]]
+    *
+    * @param iri
+    *   identifier of the resource to be resolved
+    * @param caller
+    *   user having requested the resolution
+    */
+  def resolve(
+      iri: Iri
+  )(implicit caller: Caller): IO[ElasticSearchQueryError, Result] = {
+    val locate  = ResourcesSearchParams(id = Some(iri))
+    val request = RootSearch(locate, FromPagination(0, 10000), SortList.empty)
+
+    defaultViewsQuery
+      .list(request)
+      .flatMap { searchResults =>
+        searchResults.results match {
+          case Nil         => IO.raiseError(AuthorizationFailed)
+          case Seq(result) =>
+            projectRefFromSource(result.source)
+              .traverse { projectRef =>
+                val resourceRef = ResourceRef(iri)
+                fetchResource(resourceRef, projectRef).map {
+                  _.map(SingleResult(resourceRef, projectRef, _))
+                    .getOrElse(UnexpectedError)
+                }
+              }
+              .map(_.getOrElse(UnexpectedError))
+          case _           => UIO.pure(MultipleResults(searchResults))
+        }
+      }
+  }
+
+  private def projectRefFromSource(source: JsonObject) =
+    source("_project")
+      .flatMap(_.as[Iri].toOption)
+      .flatMap(projectRefFromIri)
+
+  private val projectRefRegex =
+    s"^.+/projects/(${Label.regex.regex})/(${Label.regex.regex})".r
+
+  private def projectRefFromIri(iri: Iri) =
+    iri.toString match {
+      case projectRefRegex(org, proj) =>
+        Some(ProjectRef(Label.unsafe(org), Label.unsafe(proj)))
+      case _                          =>
+        None
+    }
+}
+
+object IdResolutionResponse {
+  sealed trait Result
+
+  sealed trait Error extends Result {
+    def reason: String
+  }
+
+  final case object UnexpectedError extends Error {
```

Comment:
Has it should not happen you can make it a terminal error for Monix, so no need to have an Encoder for it for example
It still need a message though